src/log.rs

```diff
@@ -0,0 +1,442 @@
+            }
+            Err(_) => {
+                // should we log something here?
+            }
+        }
+    }
+    let func = Some(Box_::new(Box_::new(func)));
+    match PRINT_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_print_handler(Some(func_func as _)) };
+}
+
+/// To set the default print handler, use the [`set_print_handler`] function.
+pub fn unset_print_handler() {
+    match PRINT_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = None;
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_print_handler(None) };
+}
+
+static PRINTERR_HANDLER: Lazy<Mutex<Option<Box_<Box_<dyn Fn(&str) + Send + Sync + 'static>>>>> =
+    Lazy::new(|| Mutex::new(None));
+
+/// To set back the default print handler, use the [`unset_printerr_handler`] function.
+pub fn set_printerr_handler<P: Fn(&str) + Send + Sync + 'static>(func: P) {
+    unsafe extern "C" fn func_func(string: *const libc::c_char) {
+        match PRINTERR_HANDLER.lock() {
+            Ok(handler) => {
+                if let Some(ref handler) = *handler {
+                    let string: GString = from_glib_borrow(string);
+                    (*handler)(string.as_str())
+                } else {
+                    panic!("PRINTERR_HANDLER cannot be None!");
+                }
+            }
+            Err(_) => {
+                // should we log something here?
+            }
+        }
+    }
+    let func = Some(Box_::new(Box_::new(func)));
+    match PRINTERR_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = unsafe { ::std::mem::transmute(func) };
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(Some(func_func as _)) };
+}
+
+/// To set the default print handler, use the [`set_printerr_handler`] function.
+pub fn unset_printerr_handler() {
+    match PRINTERR_HANDLER.lock() {
+        Ok(mut handler) => {
+            *handler = None;
+        }
+        Err(_) => {
+            // should we log something?
+        }
+    }
+    unsafe { glib_sys::g_set_printerr_handler(None) };
+}
+
+static DEFAULT_HANDLER: Lazy<
+    Mutex<Option<Box_<Box_<dyn Fn(&str, LogLevels, &str) + Send + Sync + 'static>>>>,
+> = Lazy::new(|| Mutex::new(None));
+
+/// To set back the default print handler, use the [`log_unset_default_handler`] function.
+pub fn log_set_default_handler<P: Fn(&str, LogLevels, &str) + Send + Sync + 'static>(log_func: P) {
+    unsafe extern "C" fn func_func(
+        log_domain: *const libc::c_char,
+        log_levels: glib_sys::GLogLevelFlags,
+        message: *const libc::c_char,
+        _user_data: glib_sys::gpointer,
+    ) {
+        match DEFAULT_HANDLER.lock() {
```

Comment:
This lock (and also elsewhere) is problematic, as you hold it while calling the function. If the function causes more logging then we'll have a deadlock.

Better to store an `Arc` inside the mutex, lock, clone the arc, unlock, then call the function.